### PR TITLE
Use (ObjectID, VersionNumber) as primary key for object storage.

### LINKED ISSUE #1318: Change object storage to use ObjectDigest or (ObjectID, version) as primary key
Currently, the objects table in SuiDataStore uses `objectID` as the primary key.    Since objects have many different versions, it means that this table only keeps the "latest" objects.    There are some issues with the current setup:

* **Writes are not idempotent**.  This is a safety issue, because it means replaying an old transaction could result in an older object overwriting a newer object.   Code must be written to ensure proper ordering of writes, which adds complication.
    * This is currently protected using a critical region and locks.  However, as the system scales out, critical regions and locks become more difficult to maintain.
    * Idempotent writes are critical for modularity and separating out components such as a lock service.  It means things may be stored in different components and batches can no longer be used everywhere.
    * Idempotent writes means ordering matters much less.
* **Locking is required to protect writes**.  See issue above.  Locks also can create performance problems, though this is not a current bottleneck.
* **objectID is not cacheable**.  Lets say we were to use a distributed object store and wanted to cache object reads.  If we use objectID by itself, one does not know the version of the object one is getting back.  This makes caching impossible because one cannot know when the underlying object changes.
* **checkpoint support**.  Supporting checkpoints will require storing multiple versions of objects.   This could be done with another table but that makes things messy, because to read a checkpoint objects one might have to span two tables then.
* Gateway stores all versions of objects currently and requires a separate table and separate code.  

The proposal here is to move to one of the following as a primary key for the objects table:
1. `(ObjectID, SequenceNumber)`
2. ObjectDigest

The primary key must be guaranteed unique for any possible new object.  It is possible that a weaker guarantee is fine for the subset of objects that lives together at the same time in one authority/gateway/etc.

When we write a new object, we would have a policy that decides whether to delete the older object. 
* Gateways would not delete the old object
* When we make a checkpoint, we would just not delete the older object.  One way of checking this is that if the current checkpoint number is higher than the checkpoint number in the object, we would just not delete it in the authority.
* Otherwise delete the old object

/cc @sblackshear @lxfind @oxade @gdanezis

### PR DESCRIPTION
**Pull request by mystenmark**
Fixes #1318 

I have some reservations about this change, however.

One of the motivations for this change was that we would only need a single table for validators and gateways/nodes, whereas we currently need one table for live objects and another for historical objects. However, it isn't so simple: Deletion of objects is represented by the removal of the object from the `objects` table. After this change, deletion necessarily involves removing all objects, not just the most recent object. This means we can't actually use a single table for historical and live objects.

We could address this by adding a `deleted` flag to the table, but this makes me doubtful that this change is still justified.

There is also the issue of write idempotency and reduced locking, which may or may not be solved by this - if we remove the critical sections in this code, will we be introducing a race condition where one transaction can delete the object versions created by another before it is finished?

### CODE REVIEW

**Comment by lxfind** [comment]
> One of the motivations for this change was that we would only need a single table for validators and gateways/nodes, whereas we currently need one table for live objects and another for historical objects. However, it isn't so simple: Deletion of objects is represented by the removal of the object from the `objects` table. After this change, deletion necessarily involves removing all objects, not just the most recent object. This means we can't actually use a single table for historical and live objects.

I think the above reason is not very important and quite minor, so let's not care about it for a moment.

> 
> There is also the issue of write idempotency and reduced locking, which may or may not be solved by this - if we remove the critical sections in this code, will we be introducing a race condition where one transaction can delete the object versions created by another before it is finished?

I assume this was the primary motivation that @velvia wanted to make this change, so let's confirm this could indeed solve the idempotency issue.

File: sui_core/src/authority/authority_store.rs [hunk 1]
```diff
@@ -240,7 +243,7 @@ impl<const ALL_OBJ_VER: bool, S: Eq + Serialize + for<'de> Deserialize<'de>>
         Ok(self
             .objects
             .iter()
-            .skip_to(&ObjectID::ZERO)?
+            .skip_to(&ObjectKey::ZERO)?
```
Comment (velvia) [inline sui_core/src/authority/authority_store.rs:246]: Not your fault but I feel this method should be called `objects_table_is_empty` because DB could potentially have other data but no objects.

File: sui_core/src/authority/authority_store.rs [hunk 2]
```diff
@@ -293,12 +296,25 @@ impl<const ALL_OBJ_VER: bool, S: Eq + Serialize + for<'de> Deserialize<'de>>
 
     /// Read an object and return it, or Err(ObjectNotFound) if the object was not found.
     pub fn get_object(&self, object_id: &ObjectID) -> Result<Option<Object>, SuiError> {
-        self.objects.get(object_id).map_err(|e| e.into())
+        let obj = self
```
Comment (velvia) [inline sui_core/src/authority/authority_store.rs:299]: Should probably modify the RustDoc to say, read the latest version of an object

File: sui_core/src/authority/authority_store.rs [hunk 3]
```diff
@@ -645,7 +663,13 @@ impl<const ALL_OBJ_VER: bool, S: Eq + Serialize + for<'de> Deserialize<'de>>
         // Delete objects.
         // Wrapped objects need to be deleted as well because we can no longer track their
         // content nor use them directly.
-        write_batch = write_batch.delete_batch(&self.objects, deleted.iter().map(|(id, _)| *id))?;
+        for key in deleted.iter() {
+            write_batch = write_batch.delete_range(
```
Comment (velvia) [inline sui_core/src/authority/authority_store.rs:667]: No, we really don't want to delete older versions.   This would destroy any older checkpoints or older versions intentionally kept for other reasons.  Instead, just delete the version in `deleted`, which are items which transactions intentionally want to delete:

    write_batch.delete_batch(&self.objects,
        deleted.iter().map(|(_, obj)| (obj.id(), obj.version())))

**Review by velvia — COMMENTED** [review]
@mystenmark thanks for this.  A couple comments:

- I added a comment but we don't need to delete all versions that say `deleted`.  
- Also the `deleted` is not actually removing older versions of new objects (ie existing inputs that get replaced by newer versions). so we need to insert in something that deletes older versions of new objects.   Perhaps `temporary_store` needs to be expanded to include `objects_replaced` or something like that.    Perhaps an approximation is the subset of active_inputs which are not immutable but are owned....   or maybe both owned and shared... @lxfind can confirm this.
- As far as idempotency, yes this change would enable that.

**Comment by lxfind** [comment]
> * I added a comment but we don't need to delete all versions that say `deleted`.
> * Also the `deleted` is not actually removing older versions of new objects (ie existing inputs that get replaced by newer versions). so we need to insert in something that deletes older versions of new objects.   Perhaps `temporary_store` needs to be expanded to include `objects_replaced` or something like that.    Perhaps an approximation is the subset of active_inputs which are not immutable but are owned....   or maybe both owned and shared... @lxfind can confirm this.

I don't quite follow this comment. The current authority store doesn't keep around all past versions of an object. So when we delete an object, it needs to be gone. And the reason that @mystenmark deleted all past versions I assume is just some sort of lazy garbage collection, where if there happens to be any history versions, we just delete them here. We could have developed a separate garbage collection service to delete older versions too, if we want.
We already have `written` in temporary_store for what you described, although I don't know if that's relevant here.

> * As far as idempotency, yes this change would enable that.

Please tell us more about this.

File: sui_core/src/authority/authority_store.rs (hunk 3 above)
Comment (mystenmark) [reply sui_core/src/authority/authority_store.rs:667]: This is insufficient - we need to be able to be able to answer the question "has this object been deleted".

**Comment by mystenmark** [comment]
> And the reason that @mystenmark deleted all past versions I assume is just some sort of lazy garbage collection

Actually the reason I deleted past versions is because there is no way to tell the difference between "version N is the current live version" and "the object is deleted, version N is an old version".

We can of course fix this by either adding a `deleted` flag or having a separate `live_objects` map.

But the bigger question is what @lxfind alludes to - how will we know when it is safe to delete old versions?

**Comment by velvia** [comment]
@lxfind @mystenmark 

> I don't quite follow this comment. The current authority store doesn't keep around all past versions of an object. So when we delete an object, it needs to be gone. And the reason that @mystenmark deleted all past versions I assume is just some sort of lazy garbage collection, where if there happens to be any history versions, we just delete them here. We could have developed a separate garbage collection service to delete older versions too, if we want.

> Actually the reason I deleted past versions is because there is no way to tell the difference between "version N is the current live version" and "the object is deleted, version N is an old version".

Ok, let me step back and try to address both of these comments.  There are two issues to think about here.

The first issue has to do with preserving older versions.  Yes, right now authorities do not keep older versions of objects (but see next issue).  When I made that comment and also in the original issue (#1318 ) I alluded to the need to preserve older versions for the sake of checkpointing.   We need to preserve older checkpointed objects, in order to facilitate recovery and TX replay from checkpoints.  From that standpoint, if we kept checkpointed objects in objects table, then we should not delete all versions.    However, I realize that discussion is out of scope of this PR, so we can punt on that for now.

The second issue has to do with idempotency and replays of older certificates.  Right now anybody can replay an older certificate and there is some chance (if the certificate is not written yet) that we may re-execute the certificate.  I'd be afraid, if we just deleted all versions, there might be a race condition between the deletion and if another simultaneous certificate tried to generate a newer version of that object.  Maybe our object locking prevents that from being possible, so it's nothing to worry about.  Basically my thinking is, we know the specific version that needs to be deleted.  In fact, the `delete_object(id, version)` is the API in temporary_store, so the API was created with the intention to delete that specific version, not all versions.   To play it safe, it is safer to just delete that specific version, and to do GC later.  That allows concurrent transactions that both delete and create.

## Idempotency

Ok, to explain about how this change would satisfy the idempotency requirement.

The goal is to allow these replays of certificates to happen without impacting newer transactions.
One of the reasons we have the mutexes and critical regions, which we want to get rid of, is the scenario where an older certificate can get re-executed and cause writes to an older version of an object.  This can overwrite another transaction which just wrote a newer version of an object, which would be bad.
Allowing multiple writes to happen at different versions solves this problem.
It makes each write `idempotent`, as in that each write can happen without affecting the other.

The fact that the older version might get deleted soon doesn't matter.... the point is just to allow the writes to happen without the write of the older version affecting the newer one.    Minimizing the mutation effects and allowing transactions to proceed in parallel is key to idempotency.

**Comment by velvia** [comment]
BTW the biggest TODO for this PR @mystenmark is about deleting older versions of newer objects - which is not addressed by the `deleted` thing.  Let me explain.

`deleted` is about that we no longer want object A.

Let's say that a transaction updates Object A from version N to N+1.   In this transaction, `deleted` in temporary store would not have object A in its list.   Why?  Because in the older schema, object A just appears in `written` and the newer version of it simply overwrites the older version.

So we need to explicitly add functionality to delete the older versions of objects that get replaced.  
Basically, the set of objects in `all_inputs` that is also in `written`, but with older versions, those versions specifically need to be deleted from the objects table.

**Comment by mystenmark** [comment]
What you said makes sense. However, isn't it possible in your plan that a certificate is replayed, which re-creates an old version of a previously deleted object? Won't that appear to clients as a zombie object returning from the dead?

The crux seems to me to be that this data model can't represent an old version of a now-deleted object.

**Comment by lxfind** [comment]
@velvia there seems to be some conflicts in your explanations, or I misunderstood:
> To play it safe, it is safer to just delete that specific version, and to do GC later. That allows concurrent transactions that both delete and create.

Above suggests that we should not remove the older version from the db when a new version is available, to allow replay of older certs.

> So we need to explicitly add functionality to delete the older versions of objects that get replaced.
Basically, the set of objects in all_inputs that is also in written, but with older versions, those versions specifically need to be deleted from the objects table.

Above suggests that we should remove the older version from the db when a new version is available?

**Comment by velvia** [comment]
> What you said makes sense. However, isn't it possible in your plan that a certificate is replayed, which re-creates an old version of a previously deleted object? Won't that appear to clients as a zombie object returning from the dead?

Yes, it is true that the crux is the model assumes objects only has live objects.  We do need something to differentiate, however the window when old certificates can be replayed is small, because...

> Above suggests that we should not remove the older version from the db when a new version is available, to allow replay of older certs.

After a certificate is finished and effects are written, from that point onwards, any replay of an old TX/Cert will just result in not executing the transaction, and having the TransactionInfo sent back with the certificate and effects.

> Above suggests that we should remove the older version from the db when a new version is available?

Yes, at the end of a transaction when the new version is written.

**Comment by asonnino** [comment]
Note that shared objects rely on the authority's store  mutex as well (in a subtle way). We need to make sure that we don't execute twice the same shared-object certificate (that may come on different tasks). Note that every shared-object transaction contains at least one owned-object (the gas object).

**Comment by asonnino** [comment]
Here are a bit more details of what may happen to shared objects if the mutex in the `AuthorityStore` is removed. Let's say that a shared-object certificate over transaction TX1 is sequenced before a shared-object certificate over transaction TX2. Also TX1 != TX2 and they both update the same started object SO.

1. A user send TX1 to the authority. TX1 is sequenced and its shared object SO is locked for sequence number 1. The transaction is however not executed because of some missing dependencies.
2. A user send TX2 to the authority. TX2 is sequenced and its shared object SO is locked for sequence number 2. The transaction is however not executed because of some missing dependencies.
3. The authority gets updated and receives all missing dependencies of TX1 and TX2.
4. A user submits TX1. The authority realises it already locked (TX1, SO) for sequence number 1 and starts execution. Since TX1 is the next transaction in the schedule (it has sequence number 1), TX1 is about to be executed.
5. A race condition happens. An other user submits another copy of TX1 to the authority (through another client, thus in another tokio ask). All checks pass and this second copy also gets executed. However the first copy of TX1 (step 4) is still in flight.
6. TX2 is submitted to the authority. All check pass (it has sequence number 2) and gets executed. So is thus updated accordingly.
7. The transaction TX1 of step 4 is now executed (it already passed the checks). TX1 thus reverted the state of SO (it effetely reverted the effects of TX2) because it updated the table `objects` of `AuthorityStore`.

Note that: (i) the authority described above is honest, and (ii) nother honest authority may not experience this race condition and thus would not revert TX2. The scenario above is thus a safety violation (Sui becomes unsafety).

**Comment by mystenmark** [comment]
@asonnino It sounds like the simplest fix is to ensure that we don't start executing a tx that is already in progress. This will of course involve some kind of synchronization but as its a much simpler scenario I think it will be easy to reduce lock contention as low as we want to. What do you think?

File: sui_core/src/authority/authority_store.rs (hunk 2 above)
Comment (mystenmark) [reply sui_core/src/authority/authority_store.rs:313]: This function now respects the tombstones in parent_sync, so its semantics are unchanged (as they have to be for existing tests to pass)

**Comment by mystenmark** [comment]
PTAL - this now uses the tombstones in `parent_sync` to track deletion, and currently never deletes old versions. Eventually we can add garbage collection if necessary.

**Comment by velvia** [comment]
@mystenmark @asonnino actually I'm not sure I understand the need for Mark's proposed fix.

The scenario described by @asonnino involves TX1 overwriting the latest shared object which TX2 wrote.  This happens today before this PR because the objects table does not have a version key.  That's what this PR is exactly designed to solve! After this PR, the objects table has version in its key, thus TX1 can safely write to (SO, old version) while not corrupting the state of TX2 which wrote to (SO, new version).

Unless I don't understand something.

**Comment by asonnino** [comment]
> > @asonnino It sounds like the simplest fix is to ensure that we don't start executing a tx that is already in progress. This will of course involve some kind of synchronization but as its a much simpler scenario I think it will be easy to reduce lock contention as low as we want to. What do you think?
> 
> Fleshing out my proposal, discussed with @asonnino:
> 
> * the problem Alberto identified was that we can start processing a tx while it is already in flight on another task - so if we simply mark that another task has started that tx, we can early exit in the second racing task.
> * easiest way to do this is with a big lock table like we currently use for objects, except we index it by tx digest
> * tx digests have no hotspots, unlike object digests, so contention can be made arbitrarily low by increasing the table size. also there's only one lock per tx, instead of N locks.

This should work, let me know when the code is in so I can have a deeper look (and make sure we are not forgetting anything)

File: sui_core/src/authority/authority_store.rs [hunk 4]
```diff
@@ -303,14 +307,43 @@ impl<
...
+    // Returns true if the most recent version of the object does not correspond to a deleted
+    // version.
+    fn object_is_live(&self, object_id: &ObjectID) -> Result<bool, SuiError> {
+        let obj = self
+            .parent_sync
+            .iter()
+            .skip_prior_to(&(*object_id, SequenceNumber::MAX, ObjectDigest::MAX))?
+            .next();
+
+        Ok(obj.map(|(k, _)| k.2.is_live()).unwrap_or(false))
```
[the remark this answers is not in the document]
Comment (mystenmark) [reply sui_core/src/authority/authority_store.rs:319]: sigh.... i make this mistake about 50% of the time i do a binary search. someday i'll learn. thanks for the catch!

**Review by asonnino — COMMENTED** [review]
Ok I checked the updates and I think it should work from a shared-object point of view

File: sui_core/src/authority/authority_store.rs [hunk 5]
```diff
@@ -303,14 +307,43 @@ impl<
...
+        } else {
+            let obj = self
+                .objects
+                .iter()
+                .skip_prior_to(&ObjectKey::max_for_id(object_id))?
+                .next();
+
+            match obj {
+                Some((ObjectKey(obj_id, _), obj)) if obj_id == *object_id => Ok(Some(obj)),
+                _ => Ok(None),
```
[the remark this answers is not in the document]
Comment (mystenmark) [reply sui_core/src/authority/authority_store.rs:335]: Added an error log.

I believe I fixed the race by checking parent_sync after loading the object. This is still a race, but it's a race that could happen even the two reads here were atomic - the object could be deleted right before the first read.

I would like to leave the `object_id` signature change for another PR. It affects `check_gas` which I don't fully understand. Either the existing `check_gas` method is broken (because it might consult a newer version of the gas object which breaks idempotency), or changing it to use a fixed version would be broken (because it might consult an old version). I'd like to understand that issue better before proceeding.

File: sui_core/src/authority/authority_store.rs [hunk 6]
```diff
@@ -1074,3 +1104,28 @@ impl<const A: bool, const B: bool, S: Eq + Serialize + for<'de> Deserialize<'de>
             }))
     }
 }
+
+// The primary key type for object storage.
+#[serde_as]
+#[derive(Eq, PartialEq, Clone, Copy, PartialOrd, Ord, Hash, Serialize, Deserialize)]
+struct ObjectKey(pub ObjectID, pub VersionNumber);
```
[the remark this answers is not in the document]
Comment (mystenmark) [reply sui_core/src/authority/authority_store.rs:1111]: I originally had it in messages.rs but its really an implementation detail of the store for now, so I moved it back in here. I think having it public would be confusing since it won't be clear to people when they should use ObjectKey vs ObjectRef

File: sui_core/src/authority/authority_store.rs [hunk 7]
```diff
@@ -303,14 +307,50 @@ impl<
...
+        match obj {
+            Some((ObjectKey(obj_id, _), obj)) if obj_id == *object_id => {
+                if self.object_is_alive(&obj_id)? {
+                    Ok(Some(obj))
+                } else {
+                    Ok(None)
+                }
+            }
+            _ => {
+                error!(?object_id, "Object is missing parent_sync entry, data store is inconsistent");
```
Comment (lxfind) [inline sui_core/src/authority/authority_store.rs:341]: The error log shouldn't be here, it should be normal when an object is just not in the store. I think what you are looking for is the None above with the Some case.

File: sui_core/src/authority/authority_store.rs [hunk 8]
```diff
@@ -303,14 +307,50 @@ impl<
...
+            .iter()
+            .skip_prior_to(&ObjectKey::max_for_id(object_id))?
+            .next();
+
+        // Note that the two reads in this function are (obviously) not atomic, and the
+        // object may be deleted after we have read it. Hence we check object_is_alive last
+        // so that the write to self.parent_sync gets the last word.
+        match obj {
+            Some((ObjectKey(obj_id, _), obj)) if obj_id == *object_id => {
+                if self.object_is_alive(&obj_id)? {
```
Comment (lxfind) [inline sui_core/src/authority/authority_store.rs:334]: Please add a TODO here just so we don't forget about confirming that the data race is ok.

File: sui_core/src/authority/authority_store.rs (hunk 7 above)
Comment (mystenmark) [reply sui_core/src/authority/authority_store.rs:341]: sorry - i think the lack of sleep from the new baby is affecting me more than I realize

File: sui_core/src/authority/authority_store.rs (hunk 8 above)
Comment (mystenmark) [reply sui_core/src/authority/authority_store.rs:358]: done

File: sui_core/src/authority/authority_store.rs [hunk 9]
```diff
@@ -303,14 +307,71 @@ impl<
...
+            .next();
+
+        let obj_ref = match obj {
+            Some((obj_ref, _)) if obj_ref.0 == *object_id => Some(obj_ref),
+            _ => None,
+        };
+
+        match obj_ref {
+            None => {
+                error!(
```
Comment (lxfind) [inline sui_core/src/authority/authority_store.rs:328]: I think you want to move the error log out of the function and put it in the None case of calling this function.
Because you also want to cover the case in line 323 where the object should exist but not in the parent_sync.

Comment (mystenmark) [reply sui_core/src/authority/authority_store.rs:328]: i don't think so - the none case of the calling function is what happens when the object simply doesn't exist, which doesn't indicate an inconsistency.

File: sui_core/src/authority/authority_store.rs [hunk 10]
```diff
@@ -305,12 +309,49 @@ impl<
...
+
+        match parent_entry {
+            None => {
+                error!(
+                    ?object_id,
+                    "Object is missing parent_sync entry, data store is inconsistent"
+                );
+                Ok(None)
+            }
+            Some((obj_ref, _)) if obj_ref.2.is_alive() => Ok(Some(obj)),
```
Comment (lxfind) [inline sui_core/src/authority/authority_store.rs:343]: Might want to add `obj.version() == obj_ref.1` to be safe.

**Review by lxfind — APPROVED** [review]
LGTM

File: sui_core/src/authority/authority_store.rs (hunk 10 above)
Comment (mystenmark) [reply sui_core/src/authority/authority_store.rs:343]: I don't think so - if a new version is written after fetching the object, we could have a version mismatch here, but that shouldn't cause the object to appear deleted.
